File: rs/gfn/src/env/hypercube.rs
use crate::{
    Action, Environment, InvalidTransitionError, State, Step, Transition,
};
use ndarray::{Array, Array1, Axis};

// TODO(danj): rewards function
// TODO(danj): API
// TODO(danj): loss function
// TODO(danj): training loop

#[derive(Debug, Clone)]
pub struct HypercubeState {
    coordinate: Array1<usize>,
    n_per_dim: usize,
}

impl State for HypercubeState {
    fn is_initial(&self) -> bool {
        self.coordinate.sum() == 0
    }

    fn is_terminal(&self) -> bool {
        self.coordinate.iter().any(|x| *x == self.n_per_dim - 1)
    }
}

impl HypercubeState {
    pub fn new(n_dims: usize, n_per_dim: usize) -> Self {
        Self {
            coordinate: Array1::zeros(n_dims),
            n_per_dim,
        }
    }

    fn apply(
        &self,
        action: &HypercubeAction,
    ) -> Result<Self, InvalidTransitionError> {
        if action.is_terminal() {
            return Ok(self.clone());
        }
        let mut coordinate = self.coordinate.clone();
        coordinate[action.direction] += 1;
        if coordinate[action.direction] < self.n_per_dim {
            Ok(Self {
                coordinate,
                ..*self
            })
        } else {
            Err(InvalidTransitionError {})
        }
    }
}

#[derive(Debug, Copy, Clone)]
pub struct HypercubeAction {
    direction: usize,
    terminal: usize,
}

impl Action for HypercubeAction {
    fn is_terminal(&self) -> bool {
        self.direction == self.terminal
    }

    fn terminate(&self) -> Self {
        Self {
            direction: self.terminal,
            terminal: self.terminal,
        }
    }
}

#[derive(Clone)]
pub struct Hypercube {
    s0: HypercubeState,
    r0: f64,
}

impl Environment for Hypercube {
    type S = HypercubeState;
    type A = HypercubeAction;
    type R = f64;

    fn s0(&self) -> Self::S {
        self.s0.clone()
    }

    fn step(
        &self,
        state: &Self::S,
        action: &Self::A,
    ) -> Result<Step<Self::S, Self::A, Self::R>, InvalidTransitionError> {
        let transition = Transition {
            state: state.clone(),
            action: action.clone(),
            next_state: state.apply(action)?,
        };
        let reward = self.reward(&transition);
        Ok(Step { transition, reward })
    }

    fn sample(&self, state: &Self::S) -> Self::A {
        // TODO(danj): use model to sample state
    }

    fn reward(
        &self,
        transition: &Transition<HypercubeState, HypercubeAction>,
    ) -> f64 {
        if !transition.is_terminal() {
            return 0.0;
        }
        let s_t = transition.next_state;
        let x_abs =
            (s_t.coordinate / (s_t.n_per_dim - 1) * 2 - 1).mapv(f64::abs);
        self.r0 + 0.5
    }
}

impl Hypercube {
    pub fn new(n_dims: usize, n_per_dim: usize, r_0: f64) -> Self {
        Self {
            s0: HypercubeState::new(n_dims, n_per_dim),
            r0,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_hypercube() {
        let (n_dims, n_per_dim, r0) = (2, 8, 0.01);
        let grid = Hypercube::new(n_dims, n_per_dim, r0);
        let s0 = grid.s0();
        let a0 = grid.sample(&s0);
        let step = grid.step(&s0, &a0).expect("simple step error!");
        let r = grid.reward(&step.transition);
        assert_eq!(r, -1.0);
    }
}

File: py/tb/tb/env/hypercube.py
from dataclasses import dataclass, replace
from typing import Self

import numpy as np
import torch
from core import InvalidTransitionError, Step, Transition
from torch import Tensor
from torch.nn import functional as F


@dataclass(frozen=True)
class HypercubeState:
    coordinate: np.ndarray
    n_per_dim: int

    def is_initial(self) -> bool:
        return self.coordinate.sum() == 0

    def is_terminal(self) -> bool:
        return (self.coordinate == self.n_per_dim - 1).any()

    def apply(self, action: "HypercubeAction") -> Self:
        if action.is_terminal():
            return self
        coordinate = self.coordinate.copy()
        coordinate[action.direction] += 1
        if coordinate[action.direction] < self.n_per_dim:
            return replace(self, coordinate=coordinate)
        raise InvalidTransitionError

    def into_tensor(self) -> Tensor:
        t = torch.tensor(self.coordinate, dtype=torch.long)
        return F.one_hot(t, self.n_per_dim).flatten()


@dataclass(frozen=True)
class HypercubeAction:
    direction: int
    terminal: int

    def is_terminal(self) -> bool:
        return self.direction == self.terminal


@dataclass(frozen=True)
class Hypercube:
    n_dims: int = 2
    n_per_dim: int = 8
    r_0: float = 1e-3

    @property
    def n_states(self) -> int:
        return self.n_dims * self.n_per_dim

    @property
    def n_actions(self) -> int:
        return self.n_dims + 1

    def initial_state(self) -> HypercubeState:
        return HypercubeState(np.zeros(self.n_dims), self.n_per_dim)

    def step(
        self, state: HypercubeState, action: HypercubeAction
    ) -> Step[HypercubeState, HypercubeAction]:
        transition = Transition(state, action, state.apply(action))
        return Step(transition, self._reward(transition.next_state))

    def _reward(self, transition):
        if not transition.is_terminal():
            return 0.0
        s_t = transition.next_state
        x_abs = np.abs(s_t / (self.n_per_dim - 1) * 2 - 1)
        return (
            self.r_0
            + 0.5 * (x_abs > 0.5).prod()
            + 2 * ((x_abs > 0.6) * (x_abs < 0.8)).prod()
        )
